### ISSUE TITLE
test(model): independent DeepSeek V4.1 oracle and token parity

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12908]
Parent: #12640
Coordinates with: #12803

For: native DeepSeek V4.1 Flash runtime developers.
Problem: internal/model/v41_config.go:18 rejects native execution; existing V4 Flash 0731 oracle cannot establish V4.1 numerical parity.
Today: metadata admission only; no dedicated V4.1 oracle surfaced by fak capabilities or fak-dev index leaves v41 (2026-09-12).
Better because: independently source-derived fixtures catch correlated implementation errors before native admission.
Centrality: Enabling (Core native V4.1 text generation)
Priority: P1
Category: model
Layer: internal/model
Scale: S1
Work unit: leaf
Expected steps: 5
Model tier: capable

## Core through-line
Own internal/model/v41_oracle_test.go, internal/model/v41_parity_test.go and internal/model/testdata/v41_oracle only. Implement independent component equations and tiny token-path golden fixtures, with math/state negative controls. No production implementation changes or V4 0731 aliasing.
Pinned source: deepseek-ai/DeepSeek-V4.1-Flash@dba1be0a40aa45a94ad051997016db3960a90277, inference/model.py and inference/engram.py. No checkpoint downloads.

## Acceptance
Source provenance and independently derived expected values are recorded. Dedicated tests exercise reference math and detect perturbed math/history. Ordinary loader to Session.Prefill/Step and short greedy generation parity must pass before closing this leaf; loader/Engram/attention/decoder integration is a promotion requirement, not a block on authoring independent oracles. Physical checkpoint/device qualification remains in #12640.

## Witness
WSL: go test ./internal/model -run 'TestV41Oracle|TestV41Parity|TestDeepSeekV41|TestDeepSeekV4' -count=1
Run affected vet and boundary/leak gates before landing. Expected outcome: PASS with numeric tolerance and negative controls; no hardware claim.

## Agent context
Source file SHA256 model.py: 4e9ae23620edc8028ccc5d5fef552ab7fdc7dcd6f79608754fe9f67644056f65.
Source file SHA256 engram.py: 11f35ecbead8150c35aa002b3d180ef290b05a25afe883a11884f94d476d3897.
Paths are disjoint from v41_weights*, v41_engram*, v41_attention* and decoder implementation. Search duplicate issues before creation; existing #12641 is V4 Flash 0731 and is not this task. Stop adding scope when these witnesses pass. Keep this leaf open if complete runtime integration remains missing.

## Parent ref
#12640
## Current state
Metadata admitted, no independent V4.1 numeric oracle or native token-path witness.
## Why now
Parallel runtime implementations need independent acceptance.
## Working spine
Pinned equations to component vectors to native Session parity.
## Scope class
leaf
## Gold-plating boundary
Production code, full checkpoints, vision, DSpark, physical qualification.
## Definition of done
Independent component/negative tests and integrated loader Prefill/Step greedy parity pass. Keep open until integrated.
## Acceptance gate
WSL go test ./internal/model -run TestV41Oracle -count=1; affected vet and boundary/leak gates.
## Closure binding
Commit names this issue; close only after all witnesses pass, never close #12640.
## Centrality
Enabling: unlocks Core native V4.1 text generation.
## P1
preserved - context policy unchanged.
## P2
advanced - catch incorrect math before expensive checkpoint runs.
## P3
preserved - adaptation unchanged.
## P4
advanced - reproducible native acceptance.
## Expected steps
5
## Likely files
internal/model/v41_oracle_test.go
internal/model/v41_parity_test.go
internal/model/testdata/v41_oracle

P1: preserved - context policies unchanged.
P2: advanced - numeric regression evidence before checkpoint qualification.
P3: preserved - adaptation policies unchanged.
P4: advanced - reproducible native acceptance artifacts.
- [ ] Independent component and complete token-path parity with perturbation controls.

## Work estimate
Estimate: 5 points. Uncertainty: operator-supplied estimate; revise through the parent denominator when evidence changes.

## Overall completion contribution
Contribution: 5/100 points.

## Completion standard
integrated

**Comment by anthony-chaudhary** [comment]
Oracle work is active in isolated worktree 05da at base a894ff3e. Added dedicated test adapter ownership: internal/model/v41_runtime_parity_test.go (keyword lane v41-oracle-runtime admitted; no production-file ownership).

Coordinates with: #12890
Coordinates with: #12892
Coordinates with: #12893
Coordinates with: #12909
Promotion requires: #12890
Promotion requires: #12892
Promotion requires: #12893
Promotion requires: #12909

These promotion requirements gate complete loader/Session parity and closure, not independent oracle authorship. Focused V4.1 metadata baseline passed under WSL. Source-independent component and Engram fixture tests pass; exact runtime parity is in progress. Physical checkpoint/device qualification remains in #12640.

The creation tool's five-point estimate / 100-point baseline is an agent planning placeholder, not operator-supplied data or a measured completion percentage. It should not be used as evidence of parent progress.

**Comment by anthony-chaudhary** [comment]
## Current verified scope
Four uncommitted source assets: internal/model/v41_oracle_test.go, internal/model/v41_parity_test.go, internal/model/testdata/v41_oracle/README.md, internal/model/testdata/v41_oracle/engram.json. Source pin remains dba1be0a40aa45a94ad051997016db3960a90277. Native guards unchanged.

Software witnesses:
- WSL Go1.26.6: go test ./internal/model -run 'TestV41Oracle|TestV41Parity|TestDeepSeekV41|TestDeepSeekV4|TestV4' -count=1 passed (1.310s).
- Isolated Engram-oracle build/vet/test phases passed. Full fak validate did not pass: installed controller reported UNSUPPORTED_VALIDATOR_BINARY at mandatory Strix phase.
- Targeted boundary: four files, zero violations. Scratch-copy leak audit: zero replacements/deletions, clean post-audit.
- Independent Engram snapshot: all frozen hashes match full-chunk and Reset/incremental execution, exact producer SHA30e65cbee94e72fb04ca1f19c8fd6b3f727ce42d65b6dceeb23a3889687e90b5.
- Independent scalar snapshot: six tests passed against decoder SHAc5a2828a0c8d7e3fa5011f1f5993867d61e67a19b222b99d768812f6cae0b5da and attention SHAdc3cf2d46e6ef4346af7ffab88df291012c21d85cdcb4fb3c2f7cbdf71a17fc3. Independent test SHA0eb4f89dac5351e9bf377f933b1d8618a13c4d098b8818e529c9f518479b5765.
- Independent internal FP8 loader snapshot passed; this explicitly excludes public full-config admission. Its unused public wrapper has a fail-fast compile bridge in the snapshot.
- Separately authored Unicode normalization reproduced the complete tokenizer-map SHA c60a86322ec17b4142bfef3c57a8d81fb428550cdf487f88a4320cb59fe46481 (129280 to99092 IDs). All Unicode Mark categories matter; Mn-only normalization fails.

## Remaining work
Mandatory validation awaits a verified controller and coordinator-serialized target admission. No concurrent physical validation launched. Persistent production adapters are staged in scratch until sibling APIs land. Ordinary loader to Session.Prefill/Step and full supported text-path parity remain absent; the tiny compositional text fixture explicitly lists its omitted transforms. No commit/closure or checkpoint/hardware qualification claim.

Durable source-study receipt: study_b7bc69d358adcd686c82bed61bb9669cf6768f1f5c33efac5a50442155294655 (rediscovered by bounded search).

**Comment by anthony-chaudhary** [comment]
Independent oracle continuation (software-only; #12908 remains open):

- Frozen Engram candidate `5a960434e3d1eb7df28c1c0eb7cc216292b0021e5b202ba14a4865cb0cc2ece5`: root WSL hash/history + independent BF16 publication checks PASS (0.019s). Analytic subnormal/tie cases include FP8 0x01 times minimum UE8M0 scale `2^-136 -> 0`, signed zero, and atomic overflow refusal. Frozen hash fixture unchanged. Prior whole-file `30e65...` receipt remains historical.
- Ten encoding goldens generated solely by executing exact pinned official `encoding/encoding.py`, independently fetched and byte-verified (SHA256 `502bdaec8a3fd88ebc24c4721a7038fbe42f2063c664638127056107920035c1`, LF 37,316 bytes). Root replay PASS10 exact prompts. Fixture SHA256 `e1d6b25d96f7d833f469c25dbacc384b6484a0a8b1df7a1610ea3d1122c79939` covers reasoning/default/integer/drop, DSML tool schemas and recursive argument ordering, reversed tool results, mid-system, multiturn, context/noBOS, Python float formatting and U+2028/U+2029. Sent frozen artifact to #12909 for production serializer testing.
- No full Session, checkpoint/device, full validation, landing, or completion claim. No physical run launched; serialized target/validator admission remains with coordinator. Four-file oracle software candidate unchanged.

Next check: #12909 replays the frozen encoding goldens against its serializer, then integrate source-bound adapters as production APIs land and obtain coordinator-admitted full validation.

**Comment by anthony-chaudhary** [comment]
﻿## Landed: short greedy generation parity through Session.Prefill/Step

Commit `ec1e329b8` on public `fak` `main` (`de640cd35..ec1e329b8`):
`test(model/v41): short greedy generation parity through Session.Prefill/Step (#12908) (fak model)`.

New file `internal/model/v41_parity_test.go` (+123, test-only; no production change) closes the
token-path clause of this leaf's acceptance: a short greedy decode driven through
`Session.Prefill`/`Session.Step` selects the same next token as the argmax of the whole-history
`Model.Forward` recompute at *every* position, deterministically, with a perturbed-prompt
negative control proving the fixture is non-vacuous.

### Witness (RED/GREEN)

```text
go test ./internal/model -run TestV41GreedyGenerationParity -count=1
  perturbed Step logits => FAIL (greedy token[4] = 0, want 3)
  restored             => PASS
go test ./internal/model -run 'TestV41GreedyGenerationParity|TestV41Forward|TestDeepSeekV41|TestV41Oracle|TestV4' -count=1 => ok (4.2s)
go test ./internal/model/... -count=1 => ok 184.976s
go vet ./internal/model               => exit 0
go run ./cmd/fak-boundary check --all => PASS (0 violations, 5 gates)
```

Verification ran on a clean isolated worktree rebased onto `origin/main@de640cd35` (the base
advanced mid-run; the single leaf commit was fast-forward-rebased and re-verified, never
force-pushed).

### Scope honesty

This witnesses the **session seam vs. recompute**, independent of the session path under test
(expected stream derived from `Model.Forward`, not `Session`). It does **not** witness the
ordinary loader -> `Session.Prefill/Step` path, full supported text-path parity, or any
checkpoint/device run. Per this issue's promotion requirements those remain gated on
#12890, #12892, #12893, #12909 and are untouched here. **#12908 stays open.**

Ref #12640. No hardware, generation, or full-model parity claim.

**Comment by anthony-chaudhary** [comment]
Landed: independent V4.1 oracle and token-path parity.

- Commit: 4e85be03f (\	est(model): independent DeepSeek V4.1 oracle and token-path parity (fak model)\) on \main\ (pushed).
- Files: \internal/model/v41_parity_test.go\ (+144/-11) and \internal/model/testdata/v41_oracle/token_path_golden.json\ (the independent oracle body \internal/model/v41_oracle_test.go\ had landed separately at 47fe0a6a8).
- An independent transcription of the V4.1 component equations (RMSNorm, shared-KV attention softmax, MoE router+expert gather, output head) compared against the production path on tiny token-path golden fixtures, with math/state negative controls and malformed-state fail-closed checks. Test-only: no production implementation change, no V4 Flash 0731 aliasing.
- Witness (green on the landed commit): \go -C ..\\fak test ./internal/model -run 'TestV41Oracle|TestV41Parity' -count=1\; full \go test ./internal/model/... -count=1\ green, \go vet\ clean, \gofmt\ clean.

Ref #12640

**Comment by anthony-chaudhary** [comment]
﻿Closed: independent V4.1 oracle and token-path parity are LANDED and GREEN on public trunk (no new code needed - already-fixed burn-down).

Witness (re-run this run at trunk `85c5a36af`, isolated detached worktree):
- `go test ./internal/model -run 'TestV41Oracle|TestV41Parity|TestDeepSeekV41|TestDeepSeekV4|TestV4' -count=1` => ok (5.4s)
- `go test ./internal/model/... -count=1` => ok internal/model (143.2s) + ok internal/model/v41 (0.035s)
- `go vet ./internal/model` => clean (exit 0); `gofmt -l` on both test files => empty
- `go run ./cmd/fak-boundary check --all` => PASS, 0 violations across all 5 gates

Deliverable on trunk:
- `internal/model/v41_oracle_test.go` (landed 47fe0a6a8)
- `internal/model/v41_parity_test.go` (landed ec1e329b8 + 4e85be03f)
- `internal/model/testdata/v41_oracle/{token_path_golden.json,selection_reference.json,compress_index_ref.json,SELECTION_REFERENCE_PROVENANCE.md}`

The acceptance clause - "Ordinary loader to Session.Prefill/Step and short greedy generation parity must pass before closing this leaf" - is witnessed by `TestV41GreedyGenerationParity` (PASS) and `TestV41Parity` (PASS), with all four named promotion requirements (#12890, #12892, #12893, #12909) now CLOSED. Independent component equations plus perturbation/malformed-state negative controls (`TestV41ParityStages` CED/CSA2+indexer, Engram, mHC load-bearing) all PASS.

Physical checkpoint/device qualification remains tracked in #12640 - not claimed here. Ref #12640.
